### PR TITLE
Fix https://github.com/zbirenbaum/copilot-cmp/issues/5

### PR DESCRIPTION
**Pull request by NOBLES5E**

### CODE REVIEW

**Comment by lysandroc** [comment]
I have made a couple tests on this, but I couldn't succeed in the tests

I'm defined [NOBLES5E:master](https://github.com/NOBLES5E/copilot-cmp) as my plugin instead of copilot-cmp. (I'm using Lazy as package manager)
image

In the cmp, I have defined this shortcut so once i press control-n it should show me copilot options
image

this is what is looks like. I was expecting copilot-cmp give me options in my cmp
image

File: lua/copilot_cmp/completion_functions.lua [hunk 1]
```diff
@@ -55,15 +55,19 @@ methods.getCompletionsCycling = function (self, params, callback)
...
+      callback(format_completions({}, params.context, self.formatters))
+      return 
+    end
     local completions = vim.tbl_values(add_results(response.completions, params))
     callback(format_completions(completions, params.context, self.formatters))
   end
 
   request("getCompletionsCycling", util.get_completion_params(), respond_callback)
-  -- Callback to cmp with empty completions so it doesn't freeze
-  callback(format_completions({}, params.context, self.formatters))
```
Comment (lysandroc) [inline lua/copilot_cmp/completion_functions.lua:66]: Just deleting this line worked for me.

File: lua/copilot_cmp/completion_functions.lua [hunk 2]
```diff
@@ -55,15 +55,19 @@ methods.getCompletionsCycling = function (self, params, callback)
...
 
   local respond_callback = function(err, response)
-    if err then return err end
-    if not response or vim.tbl_isempty(response.completions) then return end
+    if err then 
+      callback(format_completions({}, params.context, self.formatters))
+      return err 
+    end
+    if not response or vim.tbl_isempty(response.completions) then 
+      callback(format_completions({}, params.context, self.formatters))
```
Comment (lysandroc) [inline lua/copilot_cmp/completion_functions.lua:63]: `cmp` seems to continue the flow when the callback is executed, it means `cmp.completion()` got the data response and populated it in the popup list.

**Comment by zbirenbaum** [comment]
This isn't something I can merge as-is. If this allows completions to show up on an empty line, maybe I can set it up as a config flag, but the comment I have in the line removed is very important. Removing it will prevent the cmp window from appearing for any source until after a response has been obtained from copilot.

If you don't have an active internet connection, I'm pretty sure this would altogether disable completions.

Since many changes have been made that would be hard to rebase onto, you may want to update this by opening a new (draft) PR where this can be discussed. The equivalent to your code I believe would be:
```
methods.getCompletionsCycling = function (self, params, callback)
  local respond_callback = function(err, response)
    if err or not response or vim.tbl_isempty(response.completions) then
      return callback({ IsIncomplete = true, items = {}})
    end
    local completions = vim.tbl_values(add_results(response.completions, params))
    callback(format_completions(completions, params.context, self.formatters))
  end
  api.get_completions_cycling(self.client, util.get_doc_params(), respond_callback)
end

```

**Comment by gstokkink** [comment]
@zbirenbaum that works perfectly for me, perhaps that can be merged? Or are there any downsides to that approach?

**Comment by zbirenbaum** [comment]
> @zbirenbaum that works perfectly for me, perhaps that can be merged? Or are there any downsides to that approach?

There's a sizable downside in that cmp won't provide **any** completions from any source (like lsp) until copilot reponds. Copilot being http request based means it inherently is going to be the slowest completion source 99% of the time. This is not acceptable as default behavior, but could possibly be an opt in if people here want it and a better solution isn't available.

**Comment by clotodex** [comment]
Could it be configured to work like this only on a specific manually requested completion (not a cmp-config wizard so asking instead).
Szenario:
- When I want this feature to work I typically "prepped" the code already anyways, like a comment as described in the issue.
- The only completion trigger I use then is Ctrl-space (maybe with cmp we can even configure it to check if it is on a new line) - in this case I am very ok with copilot taking the longest

I don't know if this is possible though

**Pull request closed without merging**